### PR TITLE
Naive implementaton to implicitly add task information to problem reports

### PR DESCRIPTION
**Pull request by donat**
Fixes https://github.com/gradle/gradle/issues/31420

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by donat** [comment]
@bot-gradle test QFL

**Comment by donat** [comment]
@bot-gradle test QFL

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test RFFM

**Comment by donat** [comment]
@bot-gradle test RFM

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r89/ProblemProgressEventCrossVersionTest.groovy [hunk 1]
```diff
@@ -120,9 +121,12 @@ class ProblemProgressEventCrossVersionTest extends ToolingApiSpecification {
         verifyAll(problems[0]) {
             details?.details == expectedDetails
             definition.documentationLink?.url == expectedDocumentation
-            locations.size() == 2
+            locations.size() == targetVersion >= GradleVersion.version("8.12") ? 3 : 2
```
Comment (hegyibalint) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r89/ProblemProgressEventCrossVersionTest.groovy:124]: 🤔 I still think we should drop looking at the reported location sizes and assert the existence of the locations in a problem. This approach makes maintenance much heavier as we add more and more locations.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemTaskPathTracker.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+import javax.annotation.Nullable;
+
+/**
+ * This class tracks the currently executed task. The implementation is a naive workaround for capturing the currently executed task based on the assumption that a current thread can execute at most
+ * one task at a time. The correct approach would be to capture the executed tasks via build operations (see the usages of BuildOperationAncestryTracker). We invested quite some time into it only to
+ * find that introducing another ancestry tracker usage results in a significant performance regression.
+ *
+ * @see <a href="https://github.com/gradle/gradle/issues/31430">https://github.com/gradle/gradle/issues/31430</a>
+ */
```
Comment (hegyibalint) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemTaskPathTracker.java:27]: 🎉 Thanks for giving a proper reason why this exists. 
I think whoever will see this in the future will appreciate the story of why this was necessary.

File: platforms/core-execution/worker-main/src/main/java/org/gradle/process/internal/worker/problem/WorkerProblemProtocol.java [hunk 3]
```diff
@@ -26,4 +26,5 @@
  */
 public interface WorkerProblemProtocol {
     void reportProblem(Problem problem, @Nullable OperationIdentifier id);
+    void reportProblem(Problem problem, @Nullable OperationIdentifier id, String taskPath);
```
Comment (hegyibalint) [inline platforms/core-execution/worker-main/src/main/java/org/gradle/process/internal/worker/problem/WorkerProblemProtocol.java:29]: :x: The `problem` is already there: why couldn't you encode the taskPath there, and not introduce a whole new parameter?

Comment (donat) [reply platforms/core-execution/worker-main/src/main/java/org/gradle/process/internal/worker/problem/WorkerProblemProtocol.java:29]: Absolutely!

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r89/ProblemProgressEventCrossVersionTest.groovy (hunk 1 above)
Comment (donat) [reply platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r89/ProblemProgressEventCrossVersionTest.groovy]: Agreed. For starters, I've relaxed the assertion here.

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/ReceivedProblem.groovy [hunk 4]
```diff
@@ -79,6 +84,14 @@ class ReceivedProblem implements Problem {
         operationId
     }
 
+    <T> T firstLocationOfType(Class<T> type) {
+        def locations = getOriginLocations()
+        def location = locations.find { type.isInstance(it) } as T
+        assert location != null
+        assert type.isInstance(location)
+        location
+    }
```
[the remark this answers is not in the document]
Comment (donat) [reply testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/ReceivedProblem.groovy:93]: Maintenance day?

**Comment by donat** [comment]
@hegyibalint ptal

**Comment by donat** [comment]
@bot-gradle test this

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/ReceivedProblem.groovy (hunk 4 above)
[the remark this answers is not in the document]
Comment (hegyibalint) [reply testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/ReceivedProblem.groovy:93]: Hopefully

**Review by hegyibalint — APPROVED** [review]
Merge!

**Comment by donat** [comment]
@bot-gradle merge

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java [hunk 5]
```diff
@@ -149,6 +139,8 @@ public void report(Collection<? extends Problem> problems) {
      */
     @Override
     public void report(Problem problem, OperationIdentifier id) {
+        String taskPath = ProblemTaskPathTracker.getTaskIdentityPath();
+        problem = taskPath == null ? problem : problem.toBuilder(additionalDataBuilderFactory).taskPathLocation(taskPath).build();
```
Comment (wolfs) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java:143]: 🤔 Does this add the task path location as an origin location or as a context location? I thought the task path should go into the context.

**Pull request merged**